File: py/examples/counter_broadcast.py
# Mode / Broadcast
# Launch the server in #broadcast #mode to synchronize browser state across users.
# Open `/demo` in multiple browsers and watch them synchronize in realtime.
# ---
from h2o_wave import main, app, Q, ui


@app('/demo', mode='broadcast')
async def serve(q: Q):
    if not q.client.initialized:
        if q.app.count is None:
            q.app.count = 0

        q.page['example'] = ui.form_card(box='1 1 2 1', items=[
            ui.button(name='increment', label=f'Count={q.app.count}')
        ])
        q.client.initialized = True

    if q.args.increment:
        q.app.count += 1
        q.page['example'].increment.label = f'Count={q.app.count}'

    await q.page.save()

File: r/R/examples/counter_broadcast.R
library(h2owave)

serve <- function(qo)
{
    qo$app$test <- ifelse(is.null(qo$app$test),0, qo$app$test)
    ifelse(qo$args$increment == TRUE,qo$app$test <- qo$app$test + 1,qo$app$test <- 0)
    if(length(qo$args) == 0 || qo$args$increment == FALSE){
        qo$page$add_card("example",ui_form_card(
                                                box='1 1 12 10'
                                                ,items = list(
                                                              ui_button(
                                                                        name='increment'
                                                                        ,label=paste0('Count= ',qo$app$test)
                                                                        ,primary = TRUE
                                                              )
                                                )
                                                ))
    }
    else {
        qo$page$set("example","items","0","button","label",paste0("Count=",qo$app$test))
    }
    qo$page$save()
}
app("/demo",mode='broadcast')
